File: Day-11/partitioned-log-sample/python/partitioned_log.py
from __future__ import annotations

import hashlib
from dataclasses import dataclass
from typing import Dict, Generic, TypeVar

T = TypeVar("T")


@dataclass(frozen=True)
class Message(Generic[T]):
    partition: int
    offset: int
    key: str
    payload: T


class PartitionedLog(Generic[T]):
    def __init__(self, topic: str, partitions: int) -> None:
        if not topic:
            raise ValueError("topic must be non-empty")
        if partitions <= 0:
            raise ValueError("partitions must be positive")

        self.topic = topic
        self._partitions: list[list[Message[T]]] = [[] for _ in range(partitions)]
        self._committed_offsets: Dict[str, list[int]] = {}

    def produce(self, key: str, payload: T) -> Message[T]:
        partition_id = self._partition_for_key(key)
        offset = len(self._partitions[partition_id])
        message = Message(partition=partition_id, offset=offset, key=key, payload=payload)
        self._partitions[partition_id].append(message)
        return message

    def register_group(self, group_name: str) -> None:
        if group_name not in self._committed_offsets:
            self._committed_offsets[group_name] = [0] * len(self._partitions)

    def poll(self, group_name: str, max_messages: int = 1) -> list[Message[T]]:
        if max_messages <= 0:
            raise ValueError("max_messages must be positive")

        committed = self._offsets_for_group(group_name)
        cursors = committed.copy()
        batch: list[Message[T]] = []

        while len(batch) < max_messages:
            emitted = False
            for partition_id, entries in enumerate(self._partitions):
                cursor = cursors[partition_id]
                if cursor >= len(entries):
                    continue

                batch.append(entries[cursor])
                cursors[partition_id] += 1
                emitted = True

                if len(batch) >= max_messages:
                    break

            if not emitted:
                break

        return batch

    def commit(self, group_name: str, messages: list[Message[T]]) -> None:
        committed = self._offsets_for_group(group_name)
        for message in messages:
            next_offset = message.offset + 1
            if next_offset > committed[message.partition]:
                committed[message.partition] = next_offset

    def lag(self, group_name: str) -> dict[int, int]:
        committed = self._offsets_for_group(group_name)
        return {
            partition_id: len(entries) - committed[partition_id]
            for partition_id, entries in enumerate(self._partitions)
        }

    def committed_offsets(self, group_name: str) -> dict[int, int]:
        committed = self._offsets_for_group(group_name)
        return {partition_id: offset for partition_id, offset in enumerate(committed)}

    def _offsets_for_group(self, group_name: str) -> list[int]:
        offsets = self._committed_offsets.get(group_name)
        if offsets is None:
            raise KeyError(f"unknown consumer group: {group_name}")
        return offsets

    def _partition_for_key(self, key: str) -> int:
        digest = hashlib.md5(key.encode("utf-8")).hexdigest()
        return int(digest, 16) % len(self._partitions)

File: Day-11/partitioned-log-sample/java/PartitionedLog.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.math.BigInteger;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

public class PartitionedLog<T> {
    public static final class Message<T> {
        private final int partition;
        private final int offset;
        private final String key;
        private final T payload;

        private Message(int partition, int offset, String key, T payload) {
            this.partition = partition;
            this.offset = offset;
            this.key = key;
            this.payload = payload;
        }

        public int getPartition() {
            return partition;
        }

        public int getOffset() {
            return offset;
        }

        public String getKey() {
            return key;
        }

        public T getPayload() {
            return payload;
        }
    }

    private final String topic;
    private final List<List<Message<T>>> partitions;
    private final Map<String, int[]> committedOffsets;

    public PartitionedLog(String topic, int partitionCount) {
        if (topic == null || topic.isBlank()) {
            throw new IllegalArgumentException("topic must be non-empty");
        }
        if (partitionCount <= 0) {
            throw new IllegalArgumentException("partitionCount must be positive");
        }

        this.topic = topic;
        this.partitions = new ArrayList<>();
        for (int i = 0; i < partitionCount; i += 1) {
            this.partitions.add(new ArrayList<>());
        }
        this.committedOffsets = new LinkedHashMap<>();
    }

    public String getTopic() {
        return topic;
    }

    public Message<T> produce(String key, T payload) {
        int partitionId = partitionForKey(key);
        List<Message<T>> partition = partitions.get(partitionId);
        int offset = partition.size();

        Message<T> message = new Message<>(partitionId, offset, key, payload);
        partition.add(message);
        return message;
    }

    public void registerGroup(String groupName) {
        committedOffsets.computeIfAbsent(groupName, ignored -> new int[partitions.size()]);
    }

    public List<Message<T>> poll(String groupName, int maxMessages) {
        if (maxMessages <= 0) {
            throw new IllegalArgumentException("maxMessages must be positive");
        }

        int[] committed = offsetsForGroup(groupName);
        int[] cursors = Arrays.copyOf(committed, committed.length);
        List<Message<T>> batch = new ArrayList<>();

        while (batch.size() < maxMessages) {
            boolean emitted = false;

            for (int partitionId = 0; partitionId < partitions.size(); partitionId += 1) {
                List<Message<T>> entries = partitions.get(partitionId);
                int cursor = cursors[partitionId];
                if (cursor >= entries.size()) {
                    continue;
                }

                batch.add(entries.get(cursor));
                cursors[partitionId] += 1;
                emitted = true;

                if (batch.size() >= maxMessages) {
                    break;
                }
            }

            if (!emitted) {
                break;
            }
        }

        return batch;
    }

    public void commit(String groupName, List<Message<T>> messages) {
        int[] committed = offsetsForGroup(groupName);
        for (Message<T> message : messages) {
            int nextOffset = message.getOffset() + 1;
            int partitionId = message.getPartition();
            if (nextOffset > committed[partitionId]) {
                committed[partitionId] = nextOffset;
            }
        }
    }

    public Map<Integer, Integer> lag(String groupName) {
        int[] committed = offsetsForGroup(groupName);
        Map<Integer, Integer> lag = new LinkedHashMap<>();

        for (int partitionId = 0; partitionId < partitions.size(); partitionId += 1) {
            int partitionSize = partitions.get(partitionId).size();
            lag.put(partitionId, partitionSize - committed[partitionId]);
        }

        return lag;
    }

    public Map<Integer, Integer> committedOffsets(String groupName) {
        int[] committed = offsetsForGroup(groupName);
        Map<Integer, Integer> offsets = new LinkedHashMap<>();
        for (int partitionId = 0; partitionId < committed.length; partitionId += 1) {
            offsets.put(partitionId, committed[partitionId]);
        }
        return offsets;
    }

    private int[] offsetsForGroup(String groupName) {
        int[] offsets = committedOffsets.get(groupName);
        if (offsets == null) {
            throw new IllegalArgumentException("unknown consumer group: " + groupName);
        }
        return offsets;
    }

    private int partitionForKey(String key) {
        try {
            byte[] digest = MessageDigest.getInstance("MD5").digest(key.getBytes(StandardCharsets.UTF_8));
            BigInteger value = new BigInteger(1, digest);
            return value.mod(BigInteger.valueOf(partitions.size())).intValue();
        } catch (NoSuchAlgorithmException e) {
            throw new IllegalStateException("MD5 algorithm unavailable", e);
        }
    }
}
